universe/Universe.cpp

```diff
@@ -2562,8 +2562,15 @@ void Universe::UpdateEmpireStaleObjectKnowledge() {
         std::set<int>& stale_set = m_empire_stale_knowledge_object_ids[empire_id];
         const std::set<int>& destroyed_set = m_empire_known_destroyed_object_ids[empire_id];
 
-        // clear previous staleness determinations
-        stale_set.clear();
+        // remove stale marking for any known destroyed or currently visible objects
+        for (auto stale_it = stale_set.begin(); stale_it != stale_set.end();) {
+            int object_id = *stale_it;
+            if (vis_map.count(object_id) || destroyed_set.count(object_id))
+                stale_set.erase(stale_it++);
```

Comment:
I think calling erase on this iterator will invalidate it, which may be before the increment operator is applied. Need to store a copy, iterate the original, then call erase with the copy. Or use a filtering algorithm, or make a copy and iterate through that rather than the one being modified.

Reply:
Sure, good catch.  I had simply reverted that without double checking how good the quality of the original code was.

Since we are now compiling as C++11, a simpler/cleaner fix is available.